lms/djangoapps/instructor/views/api.py

```diff
@@ -726,6 +729,377 @@ def extract_user_info(user):
     return JsonResponse(response_payload)
 
 
+def _build_course_tree(course):
+    """
+    Recursively builds a course tree given the root node to a course
+    :param course tree root
+    :return list of dictionaries that represent the tree:
+    """
+    course_tree = []
+    idx = 0
+    for child in course.get_children():
+        if not child.hide_from_toc:
+            using = course.children[idx]
+            course_tree.append({
+                'display_name': child.display_name_with_default,
+                'block_id': using.block_id,
+                'block_type': using.block_type,
+                'sub': _build_course_tree(child),
+            })
+        idx += 1
+    return course_tree
+
+
+@ensure_csrf_cookie
+@cache_control(no_cache=True, no_store=True, must_revalidate=True)
+@require_level('instructor')
+def list_course_tree(request, course_id):
+    """
+    Returns a tree representing the substructures in a course
+    """
+    course_id = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    course = get_course_with_access(request.user, 'instructor', course_id, depth=None)
```

Comment:
I think having the access restricted at course staff, as in: `get_course_with_access(request.user, 'staff', course_id, depth=None)`, is better